pkg/arvo/neo/cod/std/src/fil/messenger.md

```diff
@@ -0,0 +1,649 @@
+++  kids
+  :-  some
+  :-  %y
+  %-  ~(gas by *lads:neo)
+  :~  :-  [[%.n %pub] %.n]
+      [[%pro %message-pub] (sy %sig ~)]
+      :-  [[%.n %sub] %.n]
+      [[%pro %message-sub] (sy %sig ~)]
+  ==
+++  deps  *deps:neo
+++  form
+  ...
+```
+
+The state is a `groupchat`, which is just a set of members, pending members, and a `pith` for the location of this chat in the namespace.
+
+```hoon
+$:  members=(set ship)
+    pending=(set ship)
+    host=pith
+==
+```
+
+`/imp/groupchat` takes six pokes:
+- `%invite`: Invite someone to the groupchat.
+- `%remove`: Kick someone from the groupchat. (Currently only removes their ability to post.)
+- `%invited`: Receive an invite to a groupchat.
+- `%acked`: Acknowledge acceptance of a groupchat invite.
+- `%post-to-host`: Send a post to the groupchat's host.
+- `%host-to-pub`: Send a post from the groupchat's host to their publisher shrub.
+
+```hoon
+$%  [%invite =ship provider=pith]
+    [%remove =ship]
+    [%invited host=pith]
+    [%acked ~]
+    [%post-to-host text=@t]
+    [%host-to-pub text=@t]
+==
+```
+
+#### +init
+When it's initialized, `/imp/groupchat` has either been created by a host on their own ship, or it's been created on a new member's ship by the host.
+
+If it has no state, it creates the new chat with `%message-pub` and `%message-sub` providers. If it does have some initial state, it assumes it's being created by a foreign host ship and takes that state to be the chat history. It only needs to create a `%message-sub` to receive new messages from the publisher.
+
+```hoon
+++  init
+  |=  old=(unit pail:neo)
+  ^-  (quip card:neo pail:neo)
+  ::  default case: make new groupchat with self as only member,
+    ::  and subscribe to that publisher
+  ::  XX - maybe move ordering is unpredictable here
+  ?~  old
+    :_  :-  %groupchat
+        !>([(sy our.bowl ~) ~ here.bowl])
+    :~  :-  (snoc here.bowl %pub)
+        [%make %message-pub ~ ~]
+        ::
+        :-  (snoc here.bowl %sub)
+        [%make %message-sub ~ (malt ~[[%pub (snoc here.bowl %pub)]])]
+    ==
+  ::  otherwise, I've been created as an invitee to
+  ::  someone else's groupchat
+  ?>  =(%groupchat-diff p.u.old)
+  =/  poke  !<(groupchat-diff q.u.old)
+  ?+    -.poke  !!
+      %invited
+    :_  :-  %groupchat
+        !>([~ ~ host.poke])
+    :~  :-  (snoc here.bowl %sub)
+        [%make %message-sub ~ (malt ~[[%pub (snoc host.poke %pub)]])]
+        ::
+        :-  host.poke
+        [%poke groupchat-diff/!>([%acked ~])]
+    ==
+  ==
+```
+
+#### +poke
+In the `+poke` arm we handle `%invite`, `%remove`, `%acked`, and `%post-to-host`, which is mostly a wrapper around `%host-to-pub`.
+
+Even though it doesn't handle messages — just access control — `/imp/groupchat` has some state to manage. Moreso than antyhing we've seen before, the below should look a lot like a Gall agent.
+
+{what's the heuristic for "should I store this data in "top-level" shrub state or in the namespace below it?}
```

Comment:
tl;dr INSTINCT

- Beginner-tier heuristic is "if it's a list or a set, make them as kids"
- Slightly more advanced heuristic is "who cares about this data?". Don't wanna be constantly checking your kids.bowl just to do really basic bookkeeping stuff